Implemented #62

j-hellenberg opened the pull request:
Also fixed some tests, footer is now updated on every action instead on every save.
- Discussion with customer (perfomance impact)?

Commit:
Implemented basic line number functionality

Commit:
Some refactoring; small bugfix in SquidWord building

Commit:
Fixed test of US04 (no clue why it suddenly failed, I did not touch this part of the code after all...)

Commit:
Acceptance testing for US62

Commit:
Fixed some legacy tests, added unit tests for Line/Col calculation

frcroth commented on packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st:
```diff
@@ -0,0 +1,12 @@
+
+	lineCounter := 0.
+	currentLine := self displayedLines 
+		detect: [: each | lineCounter := lineCounter + 1. each first <= (anIndex + 1) and: [each last >= (anIndex + 1)]] 
```
line too long

j-hellenberg replied:
What would correct formatting look like here?

frcroth replied:
perhaps new line before `and:`

frcroth commented on packages/RichTextEditing-Tests.package/RichTextEditorTests.class/instance/testUS04loadedTextIsInsertedIntoEditor.st:
```diff
@@ -8,4 +8,4 @@ testUS04loadedTextIsInsertedIntoEditor
 	
-	self assert: self text equals: self editorModel text
\ No newline at end of file
+	self assert: self editorModel text equals: self text 
```
trailing space

frcroth commented on packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/multipleLineText.st:
```diff
@@ -0,0 +1,11 @@
+constants
+multipleLineText
+
+	^ 'This is an text
```
Typo, should be This is **a** text

frcroth commented on packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st:
```diff
@@ -0,0 +1,7 @@
+unit-tests
+testCorrectLineColNumberAtArbitraryPositionInText
+
+	self enterStringOntoTextField: self multipleLineText.
+	self moveCaretToBeginningOfText.
+	self moveCaretMultipleTimesRight: (self standardText size + 23).
+	self assert: (self editorModel currentPositionInText) = (7@3)
```
use cascading

j-hellenberg replied:
True

Commit:
Small style fixes

oleschl commented:
Performance at least on my laptop seems fine most of the time. I tested a text with about 10000 words and performance was okay when adding new text in between old text. However, when adding new text to the end it takes more than a second for a single character. Same happens in the standard workspace.

oleschl commented on packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st:
```diff
@@ -0,0 +1,14 @@
+
+	lineCounter := 0.
+	currentLine := self displayedLines 
+		detect: 
```
Isn't this the collection count function?

j-hellenberg replied:
It actually returns the first element for which the block evaluates to true

oleschl commented on packages/RichTextEditing-Core.package/RichTextEditorModel.class/instance/positionInTextAtIndex..st:
```diff
@@ -0,0 +1,14 @@
+		ifNone: [self displayedLines last].
+	positionInLine := anIndex - currentLine first + 2.
+	
+	^ positionInLine @ lineCounter
```
I'm not sure if I like the point here or if it's better to split it into two different methods

j-hellenberg replied:
Good point, I also thought about that. Not quite sure what I prefer though

sirkrypt0 replied:
A point is fine imo, it represents the context better than two methods, as the position in the text is here really a position in form of x@y

oleschl commented:
> Also fixed some tests, footer is now updated on every action instead on every save.
> 
>     * Discussion with customer (perfomance impact)?

Kinda conflicts with the last time saved field (US86) which should be updated when saving

j-hellenberg commented:
> > Also fixed some tests, footer is now updated on every action instead on every save.
> > ```
> > * Discussion with customer (perfomance impact)?
> > ```
> 
> Kinda conflicts with the last time saved field (US86) which should be updated when saving

But we can do an footer update on saving as well, can't we? I actually have not removed it, so I don't quite see the problem

j-hellenberg commented:
> Performance at least on my laptop seems fine most of the time. I tested a text with about 10000 words and performance was okay when adding new text in between old text. However, when adding new text to the end it takes more than a second for a single character. Same happens in the standard workspace.

Writing text fast produces some lag before it catches up on my local. The normal editor also has got that problem, but not quite as much I feel. Besides that, it is working fine

sirkrypt0 commented on packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st:
```diff
@@ -0,0 +1,8 @@
+testCorrectLineColNumberAtArbitraryPositionInText
+
+	self 
+		enterStringOntoTextField: self multipleLineText;
```
You enter the `multipleLineText` here but move the caret right based on the `standardText size`. Is that intended and why?

j-hellenberg replied:
Yea, kinda. The point is, that standardText is already entered in setup but the carret is still considerered to be on position 1. The multipleLineText is appended afterwards. Therefore, I have to move it by the size of stardardText first to get an reasonable test result.

sirkrypt0 commented on packages/RichTextEditing-Tests.package/RichTextEditorTextTests.class/instance/testCorrectLineColNumberAtArbitraryPositionInText.st:
```diff
@@ -0,0 +1,8 @@
+	self 
+		enterStringOntoTextField: self multipleLineText;
+		moveCaretToBeginningOfText;
+		moveCaretMultipleTimesRight: (self standardText size + 22);
```
Where does that magic `22` come from?

oleschl committed:
Merge branch 'dev' into US62
